lib/tcpip/network_wrapper.cc

```diff
@@ -590,6 +689,128 @@ SObj network_socket_create_and_bind(Timeout       *timeout,
 }
 
+SObj network_socket_accept_tcp(Timeout        *timeout,
+                               SObj            mallocCapability,
+                               SObj            sealedListeningSocket,
+                               NetworkAddress *address,
+                               uint16_t       *port)
+{
+	SObj socket = nullptr;
+	with_sealed_socket(
+	  [&](SealedSocket *listeningSocket) {
+		  if (!check_timeout_pointer(timeout))
+		  {
+			  return -EINVAL;
+		  }
+
+		  // Create a sealed socket wrapper for the accepted connection.
+		  auto [socketWrapper, sealedSocket] = token_allocate<SealedSocket>(
+		    timeout, mallocCapability, socket_key());
+		  if (socketWrapper == nullptr)
+		  {
+			  Debug::log("Failed to allocate socket wrapper.");
+			  return -EINVAL;
+		  }
+
+		  socketWrapper->socketEpoch = currentSocketEpoch.load();
+
+		  struct freertos_sockaddr addressTmp;
+		  uint32_t                 addressLength = sizeof(addressTmp);
+		  auto                     rawSocket     = FreeRTOS_accept(
+		                            listeningSocket->socket, &addressTmp, &addressLength);
+		  if (rawSocket == nullptr)
+		  {
+			  Debug::log("Failed to create socket.");
+			  // This cannot fail unless buggy - we know that we
+			  // successfully allocated the token with this malloc
+			  // capability. Same for other calls to `token_obj_destroy`
+			  // in this function.
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+		  socketWrapper->socket = rawSocket;
+
+		  // Claim the socket so that it counts towards the caller's quota.  The
+		  // network stack also keeps a claim to it.  We will drop this claim on
+		  // deallocation.
+		  Claim c{mallocCapability, rawSocket};
+		  if (!c)
+		  {
+			  Debug::log("Failed to claim socket.");
+			  // Note that `close_socket_retry` can fail, in which case we
+			  // will leak the socket allocation. There is nothing we can do
+			  // here.  Returning the half-baked sealed socket would be
+			  // dangerous because the close() path isn't designed to handle
+			  // it (and changing it to do so is non-trivial and likely not
+			  // worth the trouble).
+			  close_socket_retry(timeout, rawSocket);
+			  // Cannot fail, see above.
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+
+		  if (LockGuard g{sealedSocketsListLock, timeout})
+		  {
+			  // Add the socket to the sealed socket reset list.
+			  socketWrapper->ring.cell_reset();
+			  sealedSockets.append(&(socketWrapper->ring));
+		  }
+		  else
+		  {
+			  // See above comments.
+			  Debug::log("Failed to add socket to the socket reset list.");
+			  close_socket_retry(timeout, rawSocket);
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+
+		  // Set `address`.
+		  if ((heap_claim_fast(timeout, address) < 0) ||
```

Comment:
You shouldn't need to claim a timeout, since check_timeout checks that it's on the stack.

Reply:
I am confused - this code is doing a claim fast on the `address`, not the timeout?